handler/handler.go

```diff
@@ -4,18 +4,63 @@ import (
 	"encoding/json"
 	"fmt"
 	"log"
+	"sync"
+	"time"
 
 	"github.com/nats-io/go-nats-streaming"
 	"github.com/openfaas/faas/gateway/queue"
 )
 
 // NatsQueue queue for work
 type NatsQueue struct {
-	nc        stan.Conn
-	ClientID  string
-	ClusterID string
-	NATSURL   string
-	Topic     string
+	nc                    stan.Conn
+	ncMutex               *sync.RWMutex
+	maxReconnect          int
+	delayBetweenReconnect time.Duration
+
+	ClientID              string
+	ClusterID             string
+	NATSURL               string
+	Topic                 string
+}
+
+func (q *NatsQueue) connect() error {
+	nc, err := stan.Connect(
+		q.ClusterID,
+		q.ClientID,
+		stan.NatsURL(q.NATSURL),
+		stan.SetConnectionLostHandler(func(conn stan.Conn, err error) {
+			log.Printf("Disconnected from %s\n", q.NATSURL)
+
+			q.reconnect(0)
+		}),
+	)
+
+	if err != nil {
+		return err
+	}
+
+	q.ncMutex.Lock()
+	q.nc = nc
+	q.ncMutex.Unlock()
+
+	return nil
+}
+
+func (q *NatsQueue) reconnect(iteration int) {
+	if iteration < q.maxReconnect {
+		time.Sleep(time.Second * time.Duration(iteration) * q.delayBetweenReconnect)
+
+		if err := q.connect(); err != nil {
+			log.Printf("Reconnection (%d) to %s failed", iteration, q.NATSURL)
+
+			q.reconnect(iteration + 1)
```

Comment:
Just curious. What is the reason for using recursion here over a bounded `for` loop?

Reply:
Had a quick chat with Ivan on NATS Slack in the #general channel, there doesn't seem to be a technical reason for the recursion. I'd prefer a loop as I think it'd be simpler. Does that work for you?

Reply:
Also, not sure if you have a way of shutting down that object (does not look like it), but if you do or planning to add, I would recommend against a sleep, especially with a backoff delay like here, since this could delay a lot the shutdown. I usually then use a select/case/time.After with selecting on a channel that can be closed in the shutdown.
Again, just a comment since there does not seem to be the case that you can shutdown...

Reply:
@alexellis do you want me to implement `shutting down logic`?

Reply:
This would be if the container received SIGINT and needs to disconnect without attempting to reconnect N times. I would suggest adding it after this PR is merged.

Reply:
👍

Reply:
There is an issue for that. I'll assign myself to it: https://github.com/openfaas/nats-queue-worker/issues/51